pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -895,16 +904,27 @@ public void redeliverUnacknowledgedMessages() {
     }
 
     @Override
-    public void redeliverUnacknowledgedMessages(List<MessageIdImpl> messageIds) {
+    public void redeliverUnacknowledgedMessages(Set<MessageIdImpl> messageIds) {
```

Comment:
I think instead of receiving list of redelivery-messages, if it just takes number of messages to be redelivered then : 
- At #L920 instead of ```incomingMessages.remove(message)```(which seems not a good idea), we can just poll (as messages will be in sequence in a queue) that many messages from ```incomingMessages``` queue, prepare a list and send for redelivery.